### ISSUE TITLE
Web-target. XMLHttpRequest (CORS) exception on flutter web project when querying servers with self-signed certificates

### ISSUE THREAD

**Issue by pawga** [flutter/flutter#124938]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

The issue has only been identified for the web build. 
A web configuration with self-signed certificates could be used in closed enterprise segments.  And flutter web assembly for the web platform is not good now :(.

To reproduce the problem is simple:
`
docker pull pawga777/pawga_ssl_demo:0.0.1
docker run -d -p 8443:8443 pawga777/pawga_ssl_demo:0.0.1
`
Next, ask flutter client to return the answer to the simplest requests:
`
https://localhost:8443/actuator/health
`

### Expected results

To the request "https://localhost:8443/actuator/health":
`
{
    "status": "UP"
}
`

### Actual results

To the request "https://localhost:8443/actuator/health":
`
XMLHttpRequest error.
`

### Code sample

I wrote two public apps to reproduce the problem:
[flutter client](https://github.com/pawga/flutter_front_admin) and ["backendon Java Spring Boot"](https://github.com/pawga/ssl-demo.git).  The backend server is implemented by docker, which can be downloaded from the docker hub. The server code can be viewed at the link above.

request could be:
https://localhost:8443/actuator/health
https://localhost:8443/actuator/info
https://localhost:8443/
https://localhost:8443/say-hello?username=Name

<details><summary>Main code:</summary>

Variant 1:
```dart
import 'package:dio/dio.dart';

void main() async {
  final dio = Dio();
  final response = await dio.get('https://localhost:8443/actuator/health');
  print(response.data);
}
```

Variant 2:
```dart
import 'dart:convert' as convert;

import 'package:http/http.dart' as http;

void main(List<String> arguments) async {

  var url =
  Uri.https('localhost:8443', 'actuator/health');

  var response = await http.get(url);
  if (response.statusCode == 200) {
    var jsonResponse =
    convert.jsonDecode(response.body) as Map<String, dynamic>;
    var itemCount = jsonResponse['totalItems'];
    print('Number of books about http: $itemCount.');
  } else {
    print('Request failed with status: ${response.statusCode}.');
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots </summary>

For the flutter client, the README.md also contains scripts and screenshots of the program demonstrating the error. For example: 
![1](https://user-images.githubusercontent.com/11136691/232305566-03e6cce8-af98-40f3-ae3e-aaad2ee18913.png)

![Снимок экрана от 2023-04-16 11-37-49](https://user-images.githubusercontent.com/11136691/232305570-b804066c-a94e-4b36-9f63-733f2bd41201.png)
![Снимок экрана от 2023-04-16 11-38-53](https://user-images.githubusercontent.com/11136691/232305576-24065b22-7b16-481b-ba08-
![Снимок экрана от 2023-04-16 11-39-40](https://user-images.githubusercontent.com/11136691/232305586-c2eeb916-8aec-420d-b568-3d25e3888702.png)
![Снимок экрана от 2023-04-16 11-40-57](https://user-images.githubusercontent.com/11136691/232305595-8e825ce0-6af2-4ee4-82a5-bf1bd5d01468.png)

</details>

### Logs

<details><summary>Logs</summary>

```console

Unhandled exception:
DioError [unknown]: null
Error: HandshakeException: Handshake error in client (OS Error: 
	CERTIFICATE_VERIFY_FAILED: self signed certificate(handshake.cc:393))
#0      DioMixin.fetch.<anonymous closure> (package:dio/src/dio_mixin.dart:532:7)
#1      _RootZone.runUnary (dart:async/zone.dart:1660:54)
#2      _FutureListener.handleError (dart:async/future_impl.dart:165:22)
#3      Future._propagateToListeners.handleError (dart:async/future_impl.dart:779:47)
#4      Future._propagateToListeners (dart:async/future_impl.dart:800:13)
#5      Future._completeError (dart:async/future_impl.dart:575:5)
#6      _SyncCompleter._completeError (dart:async/future_impl.dart:51:12)
#7      _Completer.completeError (dart:async/future_impl.dart:23:5)
#8      Future.any.onError (dart:async/future.dart:617:45)
#9      _RootZone.runBinary (dart:async/zone.dart:1665:54)
#10     _FutureListener.handleError (dart:async/future_impl.dart:162:22)
#11     Future._propagateToListeners.handleError (dart:async/future_impl.dart:779:47)
#12     Future._propagateToListeners (dart:async/future_impl.dart:800:13)
#13     Future._completeError (dart:async/future_impl.dart:575:5)
#14     Future._asyncCompleteError.<anonymous closure> (dart:async/future_impl.dart:666:7)
#15     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#16     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)
#17     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:123:13)
#18     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:405:11)
#19     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:429:5)
#20     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:192:26)

Process finished with exit code 255

Variant 2:
Unhandled exception:
HandshakeException: Handshake error in client (OS Error: 
	CERTIFICATE_VERIFY_FAILED: self signed certificate(handshake.cc:393))
#0      _SecureFilterImpl._handshake (dart:io-patch/secure_socket_patch.dart:99:46)
#1      _SecureFilterImpl.handshake (dart:io-patch/secure_socket_patch.dart:142:25)
#2      _RawSecureSocket._secureHandshake (dart:io/secure_socket.dart:915:54)
#3      _RawSecureSocket._tryFilter (dart:io/secure_socket.dart:1044:19)
<asynchronous suspension>


```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.7.11, on Linux Mint 20.1 5.4.0-146-generic, locale ru_RU.UTF-8)
    • Flutter version 3.7.11 on channel stable at /home/sivannikov/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f72efea43c (5 дней назад), 2023-04-11 11:57:21 -0700
    • Engine revision 1a65d409c7
    • Dart version 2.19.6
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /home/sivannikov/Android/Sdk
    • Platform android-33, build-tools 33.0.1
    • Java binary at: /home/sivannikov/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2022.2)
    • Android Studio at /home/sivannikov/android-studio
    • Flutter plugin version 73.0.2
    • Dart plugin version 222.4560
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
    • IntelliJ at /home/sivannikov/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.8109.175
    • Flutter plugin version 73.0.4
    • Dart plugin version 222.4459.16

[✓] IntelliJ IDEA Ultimate Edition (version 2022.3)
    • IntelliJ at /home/sivannikov/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/223.8836.41
    • Flutter plugin version 72.1.4
    • Dart plugin version 222.4459.16

[✓] VS Code (version 1.77.3)
    • VS Code at /usr/share/code
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Linux Mint 20.1 5.4.0-146-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 112.0.5615.121

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

```

</details>

**Comment by pawga** [comment]
There is a link to the project in the description of the request. You can switch between three ways to implement network requests: 
1) [dio](https://pub.dev/packages/dio) 
2) [http](https://pub.dev/packages/http) 
3) [http_pro](https://pub.dev/packages/http_pro). 
Flutter's behavior is identical.

**Comment by darshankawar** [comment]
Thanks for the update.

> Also since you are on Linux, check if this could also be one of the reasons for getting the reported error: #113594

Were you able to look at this issue and see if it could be related to your case which is also throwing a similar `CERTIFICATE_VERIFY_FAILED: self signed certificate in certificate` ?

**Comment by pawga** [comment]
We collect the entire application in docker containers and deliver it to a productive environment. In a production environment "good" certificates are used. On the development environment we use self-signed certificates while testing fresh functionality. The "DART_VM_OPTIONS" option is not relevant to this issue because the services talk to each other directly, without a proxy.
In my opinion, the problem is not in a specific self-signed certificate. This problem needs to be solved systematically in the flutter kernel. The developer should be able to "force" the fact of trust in a self-signed certificate. How can this be done in the curl utility using the -k or --insecure flags.Or for example in Java-kotlin I can do this:

`@Bean
    @Throws(Exception::class)
    fun disableSSLValidation(): Boolean {
        val sslContext = SSLContext.getInstance("TLS")

        sslContext.init(null, arrayOf<TrustManager>(object : X509TrustManager {
            @Throws(CertificateException::class)
            override fun checkClientTrusted(x509Certificates: Array<X509Certificate>, s: String) {
            }

            @Throws(CertificateException::class)
            override fun checkServerTrusted(x509Certificates: Array<X509Certificate>, s: String) {
            }

            override fun getAcceptedIssuers(): Array<X509Certificate?> {
                return arrayOfNulls(0)
            }
        }), null)
        HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.socketFactory)
        HttpsURLConnection.setDefaultHostnameVerifier { _, _ -> true }
        return true
    }`

**Comment by darshankawar** [comment]
Thanks for the update. I'll keep this issue open for team's input on current vs expected behavior.

/cc @yjbanov

**Comment by pawga** [comment]
If the experts find a solution to the problem, then all the developers of dart/flutter will say a big thank you. I alone could not find a solution. I analyzed many options, but there is no systemic solution. There is no solution at all for web assembly. This problem lives in different combinations and different developers constantly face it.
Thank you!

**Comment by biggs0125** [comment]
This seems like an issue for the Flutter framework itself rather than for Dart. Please remove the `dependency: dart` label or provide more info if you think the Dart team needs to look at this.

**Comment by pawga** [comment]
How can I remove a dependency that I didn't install :(. This dependency was installed by @darshankawar. My opinion is that Dart experts should also consider this problem. The network stack is Dart first of all, and then Flutter.

**Comment by nshahan** [comment]
@yjbanov This issue is appearing in the dart web compilers triage list because of the two labels "platform-web" and "dependency: dart". It doesn't look like you need anything from us at this time though. Are there any other labels that better describe the owners that could help with this issue?

**Comment by pawga** [comment]
As I described above in other programming languages and platforms, I can work with self-signed certificates. And I can do it easily programmatically. I studied similar queries on dart and flutter. The problem is still not solved on the web for anyone. For myself, I made a sad conclusion that I will not be able to make applications on Dart / Flutter for the web target. Do we all want Dart and Flutter to be unusable for web applications? 

I found a workaround for myself personally - to use trusted certificates even for development.
